macauff/matching.py

```diff
@@ -139,9 +218,150 @@ def __init__(self, joint_file_path, cat_a_file_path, cat_b_file_path):
+                            if self.end_time - datetime.datetime.now() < self.end_within:
+                                print('Rank {}: reaching job walltime. Cancelling all further work. {} chunks remain unprocessed.'
+                                      .format(self.rank, self.num_chunks_to_process))
+                                out_of_walltime = True
+                                self.chunk_queue.clear()
+                                # Blank end time so we don't re-enter polling loop
+                                self.end_time = None
+                                break
+                            if req.Get_status():
+                                break
+                            sleep(self.polling_rate)
+                        # Request complete, extract data
+                        (signal, worker_id, chunk_id) = req.wait()
+
+                    # Record completed chunk
+                    if signal == self.worker_signals['WORK_COMPLETE'] \
+                    and self.resume_file != None:
+                        print('Rank {}: chunk {} processed by rank {}. Adding to resume file.'
+                              .format(self.rank, chunk_id, worker_id))
+                        self.resume_file.write(chunk_id+'\n')
+                        # Do not buffer. Immediately commit to disk for
+                        # resilience against crashes and overrunning walltime
+                        # flush() alone is not enough. See:
+                        # https://docs.python.org/3/library/os.html#os.fsync
+                        self.resume_file.flush()
+                        os.fsync(self.resume_file)
+                        # Update number of remaining chunks to process
+                        self.num_chunks_to_process -= 1
+                    # Handle failed chunk
+                    elif signal == self.worker_signals['WORK_ERROR']:
+                        print('Rank {}: rank {} failed to process chunk {}.'
+                              .format(self.rank, worker_id, chunk_id))
+
+                    # Assign chunks until no more work.
+                    # Then count "no more work" signals until no more workers.
+                    try:
+                        new_chunk = self.chunk_queue.pop(0)
+                        signal = self.worker_signals['WORK_REQUEST']
+                        print('Rank {}: sending rank {} chunk {}'.format(self.rank, worker_id, new_chunk[0]))
+                    except IndexError:
+                        new_chunk = None
+                        signal = self.worker_signals['NO_MORE_WORK']
+                        active_workers -= 1
+
+                    sys.stdout.flush()
+
+                    self.comm.send((signal, new_chunk), dest=worker_id)
+
+                # Completed chunk processing loop. Perform post-processing if all
+                # chunks completed successfully and we are not out of walltime
+                # TODO: parallelise this?
```

Comment:
Following our non-github discussion previously, if you think that this step makes more sense inside the (MPI-enabled) loop for efficiency reasons then it may be worth moving it at this stage, as our default assumption is that the post-processing will (hopefully) be completely isolated on a per-chunk basis.

In any case, `self._postprocess_chunk()` needs calling within the serial-only loop as well as the MPI version, so if it's being moved to within the `self._process_chunk()` stage that solves the problem either way, or it needs adding above the serial/MPI `else` line.

Reply:
I've moved `self._postprocess_chunk()` into the main loop, as discussed. It should now be called with both MPI and serial versions as expected.